File: backend/app/api/v1/solicitudes.py
"""CRUD endpoints for solicitudes a Compras (reemplazo del Google Form)."""

from datetime import datetime, timezone

from fastapi import APIRouter, Depends, File, HTTPException, UploadFile, status
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.api.deps import get_ai, get_current_user, get_user_gmail, resolver_duenio
from app.core.exceptions import NotFoundError
from app.db.models.oportunidades import _PREVIOS_A_COTIZAR, EstadoOportunidad, Oportunidad
from app.db.models.respuestas_compras import RespuestaCompras
from app.db.models.solicitudes_compras import EstadoSolicitud, SolicitudCompras
from app.db.models.usuarios import Usuario
from app.db.session import get_db
from app.integrations.ai.base import AIProvider
from app.schemas.solicitud import (
    ParseRespuestaRequest,
    RespuestaComprasRead,
    SolicitudCreate,
    SolicitudDetail,
    SolicitudRead,
    SolicitudUpdate,
)
from app.services.grupos_compras import resolver_destino
from app.services.solicitudes import (
    build_email_preview,
    enviar_a_compras,
    guardar_adjuntos_solicitud,
)

router = APIRouter(prefix="/solicitudes", tags=["solicitudes"])

# Relaciones que exponen los schemas (evita N+1).
_RELATIONS = (
    selectinload(SolicitudCompras.oportunidad).selectinload(Oportunidad.cliente),
    selectinload(SolicitudCompras.solicitante),
    selectinload(SolicitudCompras.respuestas),
)


def _get_loaded(db: Session, solicitud_id: int) -> SolicitudCompras:
    solicitud = db.get(SolicitudCompras, solicitud_id, options=list(_RELATIONS))
    if solicitud is None:
        raise NotFoundError("Solicitud no encontrada")
    return solicitud


def _assert_owner(solicitud: SolicitudCompras, user: Usuario) -> None:
    """Acceso compartido: cualquier usuario puede ver/actuar sobre cualquier
    solicitud. Se mantiene (no-op) para no tocar los call sites."""
    return


def _normalize_ccs(data: dict) -> None:
    """ccs_extra llega como list[EmailStr]; la columna ARRAY espera str planos."""
    if data.get("ccs_extra"):
        data["ccs_extra"] = [str(e) for e in data["ccs_extra"]]


@router.get("", response_model=list[SolicitudRead])
def list_solicitudes(
    estado: EstadoSolicitud | None = None,
    oportunidad_id: int | None = None,
    usuario_id: int | None = None,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
) -> list[SolicitudCompras]:
    """Solicitudes personales: cada usuario ve solo las que él pidió a Compras.
    Los admin ven las de todo el equipo, o filtran por `usuario_id` (perfil)."""
    query = select(SolicitudCompras).options(*_RELATIONS)
    if estado is not None:
        query = query.where(SolicitudCompras.estado == estado)
    if oportunidad_id is not None:
        query = query.where(SolicitudCompras.oportunidad_id == oportunidad_id)
    duenio_id = resolver_duenio(current_user, usuario_id)
    if duenio_id is not None:
        query = query.where(SolicitudCompras.solicitante_id == duenio_id)
    return list(db.scalars(query.order_by(SolicitudCompras.created_at.desc())))


@router.post("", response_model=SolicitudRead, status_code=201)
def create_solicitud(
    body: SolicitudCreate,
    db: Session = Depends(get_db),
    user: Usuario = Depends(get_current_user),
) -> SolicitudCompras:
    oportunidad = db.get(Oportunidad, body.oportunidad_id)
    if oportunidad is None:
        raise NotFoundError("Oportunidad no encontrada")

    data = body.model_dump()
    _normalize_ccs(data)
    # Destino de Compras: grupo elegido (o el default del usuario). Se snapshotea
    # en la solicitud para no depender de que el grupo siga existiendo luego.
    grupo_id = data.pop("grupo_compras_id", None)
    destino_to, destino_cc = resolver_destino(db, user.id, grupo_id)
    now = datetime.now(timezone.utc)
    solicitud = SolicitudCompras(
        **data,
        solicitante_id=user.id,
        destino_to=destino_to,
        destino_cc=destino_cc or None,
        fecha_envio=now,
        estado=EstadoSolicitud.enviada,
    )
    db.add(solicitud)

    # Al solicitar a Compras, la oportunidad avanza a "en_compras".
    oportunidad.estado = EstadoOportunidad.en_compras
    oportunidad.fecha_ultimo_movimiento = now

    db.commit()
    return _get_loaded(db, solicitud.id)


@router.post("/{solicitud_id}/adjuntos", response_model=SolicitudRead)
def subir_adjuntos(
    solicitud_id: int,
    files: list[UploadFile] = File(...),
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
) -> SolicitudCompras:
    """Adjunta archivos (PDF/imágenes que mandó el cliente) a la solicitud. Se
    envían junto al mail a Compras."""
    solicitud = _get_loaded(db, solicitud_id)
    _assert_owner(solicitud, current_user)
    archivos = [
        {"filename": f.filename, "mime_type": f.content_type, "data": f.file.read()}
        for f in files
    ]
    guardar_adjuntos_solicitud(db, solicitud, archivos)
    return _get_loaded(db, solicitud_id)


@router.get("/{solicitud_id}", response_model=SolicitudDetail)
def get_solicitud(
    solicitud_id: int,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
) -> SolicitudDetail:
    solicitud = _get_loaded(db, solicitud_id)
    _assert_owner(solicitud, current_user)
    read = SolicitudRead.model_validate(solicitud)
    preview = build_email_preview(solicitud, db)
    respuestas = [RespuestaComprasRead.model_validate(r) for r in solicitud.respuestas]
    return SolicitudDetail(**read.model_dump(), email_preview=preview, respuestas=respuestas)


@router.post("/{solicitud_id}/enviar", response_model=SolicitudRead)
def enviar_solicitud(
    solicitud_id: int,
    db: Session = Depends(get_db),
    gmail=Depends(get_user_gmail),  # noqa: ANN001 - GmailClient del usuario logueado
    current_user: Usuario = Depends(get_current_user),
) -> SolicitudCompras:
    """Envía la solicitud a Compras por Gmail (desde la casilla del vendedor)."""
    solicitud = _get_loaded(db, solicitud_id)
    _assert_owner(solicitud, current_user)
    try:
        enviar_a_compras(db, gmail, solicitud)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc)) from exc
    except Exception as exc:  # noqa: BLE001 - frontera con Gmail
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=f"No se pudo enviar a Compras: {exc}",
        ) from exc
    return _get_loaded(db, solicitud_id)


@router.post(
    "/{solicitud_id}/respuesta", response_model=RespuestaComprasRead, status_code=201
)
def cargar_respuesta(
    solicitud_id: int,
    body: ParseRespuestaRequest,
    db: Session = Depends(get_db),
    ai: AIProvider = Depends(get_ai),
    current_user: Usuario = Depends(get_current_user),
) -> RespuestaCompras:
    """Parsea con IA la respuesta de Compras (tabla de precios) y la registra.

    Marca la solicitud como respondida. Los ítems quedan listos para armar el
    presupuesto (POST /presupuestos/desde-solicitud/{id})."""
    solicitud = _get_loaded(db, solicitud_id)
    _assert_owner(solicitud, current_user)
    try:
        draft = ai.draft_quote(body.contenido)
    except Exception as exc:  # noqa: BLE001 - frontera con IA
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=f"Error al parsear con IA: {exc}",
        ) from exc

    respuesta = RespuestaCompras(
        solicitud_compras_id=solicitud.id,
        contenido_raw=body.contenido,
        datos_parseados_ia=draft.model_dump(),
        notas_compras=draft.notas,
    )
    db.add(respuesta)
    ahora = datetime.now(timezone.utc)
    solicitud.estado = EstadoSolicitud.respondida
    if solicitud.fecha_respuesta is None:
        solicitud.fecha_respuesta = ahora
    # Seguimiento: registrar cuándo respondió Compras y avanzar la oportunidad
    # a "Cotizado por compras" (sin pisar estados posteriores).
    op = solicitud.oportunidad
    if op is not None:
        if op.fecha_respuesta_compras is None:
            op.fecha_respuesta_compras = ahora.date()
        if op.estado in _PREVIOS_A_COTIZAR:
            op.estado = EstadoOportunidad.cotizado_compras
            op.fecha_ultimo_movimiento = ahora
    db.commit()
    db.refresh(respuesta)
    return respuesta


@router.patch("/{solicitud_id}", response_model=SolicitudRead)
def update_solicitud(
    solicitud_id: int,
    body: SolicitudUpdate,
    db: Session = Depends(get_db),
    current_user: Usuario = Depends(get_current_user),
) -> SolicitudCompras:
    solicitud = db.get(SolicitudCompras, solicitud_id)
    if solicitud is None:
        raise NotFoundError("Solicitud no encontrada")
    _assert_owner(solicitud, current_user)

    data = body.model_dump(exclude_unset=True)
    _normalize_ccs(data)
    for field, value in data.items():
        setattr(solicitud, field, value)

    # Al marcar "respondida" registramos la fecha de respuesta si falta.
    if body.estado == EstadoSolicitud.respondida and solicitud.fecha_respuesta is None:
        solicitud.fecha_respuesta = datetime.now(timezone.utc)

    db.commit()
    return _get_loaded(db, solicitud_id)

File: frontend/lib/solicitudes.ts
// API + hooks de React Query para solicitudes a Compras.
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "@/lib/api";
import { oportunidadKeys } from "@/lib/oportunidades";
import type {
  CondicionPago,
  EstadoSolicitud,
  RespuestaCompras,
  Seccion,
  Solicitud,
  SolicitudCreate,
  SolicitudDetail,
  SolicitudUpdate,
} from "@/lib/types";

const BASE = "/api/v1/solicitudes";

export const solicitudKeys = {
  all: ["solicitudes"] as const,
  detail: (id: number) => ["solicitudes", id] as const,
};

export const CONDICIONES_PAGO: CondicionPago[] = [
  "15",
  "30",
  "45",
  "60",
  "120",
  "Transferencia",
  "Cheque Anticipado a Entrega - 15 días",
  "Cheque Anticipado a Entrega - 30 días",
  "Cheque Anticipado a Entrega - 60 días",
];

export const ESTADO_SOLICITUD_META: Record<EstadoSolicitud, { label: string; color: string }> = {
  enviada: { label: "Enviada", color: "bg-amber-100 text-amber-700" },
  respondida: { label: "Respondida", color: "bg-green-100 text-green-700" },
  cerrada: { label: "Cerrada", color: "bg-slate-100 text-slate-600" },
};

export function useSolicitudes(
  usuarioId?: number,
  oportunidadId?: number,
  ambito?: Seccion
) {
  return useQuery({
    queryKey:
      usuarioId || oportunidadId || ambito
        ? [...solicitudKeys.all, { usuarioId, oportunidadId, ambito }]
        : solicitudKeys.all,
    queryFn: async () => {
      const params: Record<string, string | number> = {};
      if (usuarioId) params.usuario_id = usuarioId;
      if (oportunidadId) params.oportunidad_id = oportunidadId;
      if (ambito) params.ambito = ambito;
      return (await api.get<Solicitud[]>(BASE, { params })).data;
    },
  });
}

export function useSolicitud(id: number | null) {
  return useQuery({
    queryKey: solicitudKeys.detail(id ?? 0),
    queryFn: async () => (await api.get<SolicitudDetail>(`${BASE}/${id}`)).data,
    enabled: id !== null && id > 0,
  });
}

// Sube los archivos adjuntos (PDF/imágenes del cliente) a una solicitud.
async function subirAdjuntos(id: number, files: File[]): Promise<void> {
  if (!files || files.length === 0) return;
  const fd = new FormData();
  files.forEach((f) => fd.append("files", f));
  await api.post(`${BASE}/${id}/adjuntos`, fd);
}

type CrearSolicitud = { body: SolicitudCreate; files?: File[] };

export function useCreateSolicitud() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ body, files }: CrearSolicitud) => {
      const solicitud = (await api.post<Solicitud>(BASE, body)).data;
      await subirAdjuntos(solicitud.id, files ?? []);
      return solicitud;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: solicitudKeys.all });
      // El alta mueve la oportunidad a "en_compras": refrescar tablero/listado.
      qc.invalidateQueries({ queryKey: oportunidadKeys.all });
    },
  });
}

// Crea la solicitud y la envía a Compras por Gmail en un solo paso (usado desde
// "Pedir a Compras" en Oportunidades). Si el envío falla, la solicitud ya quedó
// creada y se puede reintentar el envío desde /solicitudes.
export function useCrearYEnviarSolicitud() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async ({ body, files }: CrearSolicitud) => {
      const solicitud = (await api.post<Solicitud>(BASE, body)).data;
      await subirAdjuntos(solicitud.id, files ?? []);
      await api.post(`${BASE}/${solicitud.id}/enviar`);
      return solicitud;
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: solicitudKeys.all });
      qc.invalidateQueries({ queryKey: oportunidadKeys.all });
    },
  });
}

export function useUpdateSolicitud(id: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (body: SolicitudUpdate) =>
      (await api.patch<Solicitud>(`${BASE}/${id}`, body)).data,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: solicitudKeys.all });
      qc.invalidateQueries({ queryKey: solicitudKeys.detail(id) });
    },
  });
}

// Envía la solicitud a Compras por Gmail (desde la casilla del vendedor).
export function useEnviarSolicitud(id: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async () => (await api.post<Solicitud>(`${BASE}/${id}/enviar`)).data,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: solicitudKeys.detail(id) });
      qc.invalidateQueries({ queryKey: solicitudKeys.all });
    },
  });
}

// Pega el texto de la respuesta de Compras y la IA extrae los ítems.
export function useCargarRespuesta(id: number) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (contenido: string) =>
      (await api.post<RespuestaCompras>(`${BASE}/${id}/respuesta`, { contenido })).data,
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: solicitudKeys.detail(id) });
      qc.invalidateQueries({ queryKey: solicitudKeys.all });
    },
  });
}
